File: LogicAndTrick.WikiCodeParser/Nodes/RefNode.cs
using System;
using System.Collections.Generic;

namespace LogicAndTrick.WikiCodeParser.Nodes
{
    public class RefNode : INode
    {
        public ParseData Data { get; set; }
        public string Name { get; set; }

        public RefNode(ParseData data, string name)
        {
            Data = data;
            Name = name;
        }

        private INode GetNode()
        {
            return Data.Get($"Ref::{Name}", () => UnprocessablePlainTextNode.Empty);
        }

        public string ToHtml()
        {
            return GetNode().ToHtml();
        }

        public string ToPlainText()
        {
            return GetNode().ToPlainText();
        }

        public IList<INode> GetChildren()
        {
            return new List<INode>{ GetNode() };
        }

        public void ReplaceChild(int i, INode node)
        {
            if (i != 0) throw new IndexOutOfRangeException();
            Data.Set($"Ref::{Name}", node);
        }

        public bool HasContent()
        {
            return GetNode().HasContent();
        }
    }
}

File: php/src/Nodes/RefNode.php
<?php

namespace LogicAndTrick\WikiCodeParser\Nodes;

use LogicAndTrick\WikiCodeParser\ParseData;

class RefNode implements INode
{
    public ParseData $data;
    public string $name;

    public function __construct(ParseData $data, string $name)
    {
        $this->data = $data;
        $this->name = $name;
    }

    private function GetNode() : INode {
        return $this->data->Get("Ref::{$this->name}", fn() => UnprocessablePlainTextNode::Empty());
    }

    function ToHtml(): string
    {
        return $this->GetNode()->ToHtml();
    }

    function ToPlainText(): string
    {
        return $this->GetNode()->ToHtml();
    }

    function GetChildren(): array
    {
        return [$this->GetNode()];
    }

    function ReplaceChild(int $i, INode $node): void
    {
        if ($i != 0) throw new \Exception('Index out of range');
        $this->data->Set("Ref::{$this->name}", $node);
    }

    function HasContent(): bool
    {
        return $this->GetNode()->HasContent();
    }
}